woocommerce/pinterest-for-woocommerce
# Add shipping collumn.

### Changes proposed in this Pull Request:

This PR implements a shipping column for the Pinterest XML feed file. Shipping column entry is defined as a comma-separated list of shipping information in the format `country:state:method_name:cost` like for example: `US:CA:Free shipping:0.00 USD` alternatively the state may be dropped ( if not configured in the zone ) resulting in `US::Free shipping:0.00 USD`. Check the screenshot for more advanced entry.

The basic logic is to take a product and try to figure out the best shipping option available as if the product would be alone inside the cart with the shipping destination set to the one defined by the shipping zone setting ( if multiple zones and multiple locations in the zone are provided we do that for each such location.

Not all of the shipping features present in WooCommerce could be mapped onto Pinterest feed shipping column specification. List of things that are supported:

  - Free shipping without additional settings.
  - Free shipping with a minimum order value. Minimum is tested over a single product. ( still, better than nothing )
  - Flat rate shipping without additional settings.
  - Flat rate with classes: no shipping class and regular classes.
  - Shipping zone locations: single country, country + state, continent ( zip codes are not allowed).
  - Locations mixing: continent + country + state.
  - Taxes for shipping global.
  - Taxes for shipping for a specific country.
  - Filtering of not supported countries.
  - Simple products.
  - Variable products.

Right now no extensibility ( filters ) has been introduced. We can add this when we will see requests for such features.

### Screenshots:
![image](https://user-images.githubusercontent.com/17271089/150472409-d171d697-7636-4ed8-9469-f10524eaf574.png)

### Detailed test instructions:


1. Connect the plugin to Pinterest.
2. Crate some products.
3. Create some shipping zone.
4. Add shipping methods to the zones
5. Observe the feed file ( `wp-content/uploads/pinterest-for-woocommerce-xxxxxx.xml` ) and check if the `<g:shipping>` tag has an entry that makes sense given the settings that you have created.
6. Additionally add some tax settings that should impact the shipping cost and verify the output again.

### Additional details:

The PR looks big but about one-half is the testing code and not the actual implementation.

Things that we will need to observe is performance. This is a calculation heavy implementation - but given the requirements presented for this task, I have not found any shortcuts that would allow performance improvements without severely impairing the accuracy.

### Changelog entry

> Add - Product shipping information in the feed file.
